crates/bevy_pbr/src/deferred/mod.rs

```diff
@@ -0,0 +1,483 @@
+            }),
+            primitive: PrimitiveState::default(),
+            depth_stencil: Some(DepthStencilState {
+                format: DEFERRED_LIGHTING_PASS_ID_DEPTH_FORMAT,
+                depth_write_enabled: false,
+                depth_compare: CompareFunction::Equal,
+                stencil: StencilState {
+                    front: StencilFaceState::IGNORE,
+                    back: StencilFaceState::IGNORE,
+                    read_mask: 0,
+                    write_mask: 0,
+                },
+                bias: DepthBiasState {
+                    constant: 0,
+                    slope_scale: 0.0,
+                    clamp: 0.0,
+                },
+            }),
+            multisample: MultisampleState::default(),
+            push_constant_ranges: vec![],
+        }
+    }
+}
+
+impl FromWorld for DeferredLightingLayout {
+    fn from_world(world: &mut World) -> Self {
+        let render_device = world.resource::<RenderDevice>();
+        let layout = render_device.create_bind_group_layout(&BindGroupLayoutDescriptor {
+            label: Some("deferred_lighting_layout"),
+            entries: &[BindGroupLayoutEntry {
+                binding: 0,
+                visibility: ShaderStages::VERTEX_FRAGMENT,
+                ty: BindingType::Buffer {
+                    ty: render_resource::BufferBindingType::Uniform,
+                    has_dynamic_offset: false,
+                    min_binding_size: Some(PbrDeferredLightingDepthId::min_size()),
+                },
+                count: None,
+            }],
+        });
+        Self {
+            bind_group_layout_0: world.resource::<MeshPipeline>().view_layout.clone(),
+            bind_group_layout_1: layout,
+        }
+    }
+}
+
+pub fn insert_deferred_lighting_depth_id_component(
+    mut commands: Commands,
+    views: Query<Entity, (With<DeferredPrepass>, Without<PbrDeferredLightingDepthId>)>,
+) {
+    for entity in views.iter() {
+        commands
+            .entity(entity)
+            .insert(PbrDeferredLightingDepthId::default());
+    }
+}
+
+pub fn prepare_deferred_lighting_pipelines(
+    mut commands: Commands,
+    pipeline_cache: Res<PipelineCache>,
+    mut pipelines: ResMut<SpecializedRenderPipelines<DeferredLightingLayout>>,
+    differed_lighting_layout: Res<DeferredLightingLayout>,
+    views: Query<
+        (
+            Entity,
+            &ExtractedView,
+            Option<&Tonemapping>,
+            Option<&DebandDither>,
+            Option<&EnvironmentMapLight>,
+            Option<&ShadowFilteringMethod>,
+            Option<&ScreenSpaceAmbientOcclusionSettings>,
+        ),
+        With<DeferredPrepass>,
+    >,
+    images: Res<RenderAssets<Image>>,
+) {
+    for (entity, view, tonemapping, dither, environment_map, shadow_filter_method, ssao) in &views {
+        let mut view_key = MeshPipelineKey::from_hdr(view.hdr);
+
+        if !view.hdr {
+            if let Some(tonemapping) = tonemapping {
```

Comment:
so much copy/paste ...

Reply:
Agreed, @superdump mentioned moving what's common here into a separate function that's used by both in a subsequent PR. @superdump is it maybe worth doing this here or should we wait?

Reply:
I think we should wait. This PR and reviewing it are complicated enough in my opinion. :) There are some footguns around specialisation and this key gathering is definitely part of that.